pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -733,9 +733,8 @@ func removeTerminatedPods(pods []*runtimeapi.PodSandbox) []*runtimeapi.PodSandbo
 	return result
 }
 
-// removeTerminatedContainers returns containers with terminated ones.
-// It only removes a terminated container when there is a running instance
-// of the container.
+// removeTerminatedContainers removes all terminated containers since they should
+// not be used for usage calculations.
 func removeTerminatedContainers(containers []*runtimeapi.Container) []*runtimeapi.Container {
```

Comment:
I did some more digging, and I think a lot of this function can be removed.

I can only find 2 references to `removeTerminatedContainers`, and neither use the fact that the output is sorted (by container ID and by create time). Because of that, I think we can remove the whole map and sort process.

@mattjmcnaughton am I barking up the wrong tree?

Reply:
This was originally added in the `cadvisor_stats_provider` as a fix for https://github.com/kubernetes/kubernetes/issues/47853. Its primary functionality was for deduping due to a race condition between cadvisor and cgroup removal. If deduping is no longer necessary (fwiw I have not observed this race condition happening between crio and cgroups), this function can indeed be greatly simplified. I did some searching through the commit history and was unable to determine why the deduping functionality was ported forward from `cadvisor_stats_provider` when `cri_stats_provider` was added. I have preserved the deduping logic to minimize the impact of this diff, but personally I feel this can probably be removed as well.